numpy-stubs/__init__.pyi

```diff
@@ -862,21 +876,21 @@ def take(
 def reshape(a: _ArrayLike, newshape: _ShapeLike, order: _Order = ...) -> ndarray: ...
 @overload
 def choose(
-    a: _ScalarGeneric,
+    a: _ScalarInt,
     choices: Union[Sequence[_ArrayLike], ndarray],
     out: Optional[ndarray] = ...,
     mode: _Mode = ...,
-) -> _ScalarGeneric: ...
+) -> _ScalarInt: ...
 @overload
 def choose(
-    a: _Scalar,
+    a: _Int,
     choices: Union[Sequence[_ArrayLike], ndarray],
     out: Optional[ndarray] = ...,
     mode: _Mode = ...,
-) -> _ScalarNumpy: ...
+) -> Union[integer, bool_]: ...
 @overload
 def choose(
-    a: _ArrayLike,
+    a: _ArrayLikeInt,
```

Comment:
So while passing bools technically works, given that the the docstring says

"This array must contain integers in [0, n-1], where n is the number of choices"

it seems to me that this is more an accident than the intent. I'm not sure that we should allow this in the types as it seems like not a best practice. @rgommers WDYT about this case? (We could also solicit opinions on the mailing list.)

Reply:
Accident might not be the best term here.
The general compatibility between ``int`` and ``bool`` stems from the fact the ``bool`` is a subclass of the former. 

```python
>>> import numpy as np

>>> issubclass(int, bool)
True

>>> True == 1 and False == 0
True

>>> np.bool_(True) == 1 and np.bool_(False) == 0
True
```

``Union[int, bool]`` is consequently equivalent to just plain ``int``, but as ``np.bool_`` is not a subclass of ``np.integer`` this does not hold for ``np.bool_``. Even though the latter two classes behave (more or less) identically to ``int`` and ``bool``.

This also means that we _could_ remove ``np.bool_`` from the ``_ArrayLikeInt`` union, but properly removing ``bool`` would require adding a new overload along the lines of 
``choose(a: bool, ...) -> NoReturn``.

Reply:
> but as np.bool_ is not a subclass of np.integer this does not hold for np.bool_

And this is why I don't want to include it. Yes, including `int` gets you `bool`, but that's intentionally not true for `numpy.bool_`, so I'm not sure we should make an extra effort to bring it back here. To be clear, when I said "accident" I meant "that `np.bool_` works".

Reply:
Fair enough, I've implemented the name changes in https://github.com/numpy/numpy-stubs/pull/74.